chain/epoch-manager/src/epoch_sync.rs

```diff
@@ -12,12 +13,102 @@ use near_primitives::epoch_sync::{
 use tracing::instrument;
 
+/// Function to extend epoch sync proof by one epoch.
+/// - Retrieve the existing epoch sync proof from the store
+/// - Validate that the existing proof is consistent with the new epoch to be added, i.e.
+///   the proof exists upto the previous epoch.
+/// - Generate and append the new epoch data to the existing proof.
+/// - Updated proof back to the store.
+pub fn extend_epoch_sync_proof(
+    store: &EpochStoreAdapter,
+    last_block_hash: &CryptoHash,
+) -> Result<(), Error> {
+    let last_block_info = store.get_block_info(&last_block_hash)?;
+    if last_block_info.epoch_id() == &EpochId::default() {
+        // Genesis epoch, nothing to do.
+        return Ok(());
+    }
+
+    let first_block_hash = last_block_info.epoch_first_block();
+    let first_block_info = store.get_block_info(&first_block_hash)?;
+    let last_block_hash_in_prev_epoch = first_block_info.prev_hash();
+    let last_block_info_in_prev_epoch = store.get_block_info(&last_block_hash_in_prev_epoch)?;
+
+    let mut all_epochs = match store.get_epoch_sync_proof()? {
+        Some(EpochSyncProofV1 { all_epochs, current_epoch, .. }) => {
+            // Sanity check ensuring continuity of epochs. The `current_epoch` from the existing proof
+            // should correspond to the previous epoch we are adding now.
+            let expected_first_block_hash = last_block_info_in_prev_epoch.epoch_first_block();
+            let actual_first_block_hash = current_epoch.first_block_header_in_epoch.hash();
+            debug_assert_eq!(expected_first_block_hash, actual_first_block_hash);
+            all_epochs
+        }
+        None => {
+            // This should only happen if we are on the epoch after genesis.
+            debug_assert_eq!(last_block_info_in_prev_epoch.epoch_id(), &EpochId::default());
+            vec![]
+        }
+    };
+
+    all_epochs.push(get_epoch_sync_proof_epoch_data(store, &last_block_info)?);
+    let last_epoch = get_epoch_sync_proof_last_epoch_data(store, &last_block_hash_in_prev_epoch)?;
+    let current_epoch =
+        get_epoch_sync_proof_current_epoch_data(&store.chain_store(), &last_block_info)?;
+
+    let proof = EpochSyncProofV1 { all_epochs, last_epoch, current_epoch };
+    let proof = EpochSyncProof::V1(proof);
+    let mut store_update = store.store_update();
+    store_update.set_epoch_sync_proof(&proof);
+    store_update.commit()?;
+
+    Ok(())
+}
+
+fn get_epoch_sync_proof_epoch_data(
+    store: &EpochStoreAdapter,
+    last_block_info: &BlockInfo,
+) -> Result<EpochSyncProofEpochData, Error> {
+    let chain_store = store.chain_store();
+    let epoch_info = store.get_epoch_info(last_block_info.epoch_id())?;
+
+    let block_producers = get_epoch_info_block_producers(&epoch_info);
+
+    // Ideally we should we calling should_use_versioned_bp_hash_format(prev_epoch.protocol_version)
+    // to get use_versioned_bp_hash_format, but we are guaranteed that
```

Comment:
Does this mean this specific function is only ever used to extend the proof? Maybe reflect that on the name?

Reply:
Hmm, point... Yeah, it's used only to extend the proof. I'll specify that